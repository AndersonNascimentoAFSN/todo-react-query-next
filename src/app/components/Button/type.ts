import { ButtonHTMLAttributes, ReactNode } from "react"

export type VariantButtonProps = 'primary' | 'secondary' | 'tertiary'

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  children: ReactNode
  variant?: VariantButtonProps
}