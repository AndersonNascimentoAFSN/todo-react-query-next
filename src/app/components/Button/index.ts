export { Button } from './Button'
export type { ButtonProps, VariantButtonProps } from './type'