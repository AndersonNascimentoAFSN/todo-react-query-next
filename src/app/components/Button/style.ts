import { VariantButtonProps } from './type'

export const BASE_BUTTON_CLASSES =
'py-3.5 px-8 rounded-md border-2 group flex gap-3 justify-center items-center';

export const getVariantClasses = (variant: VariantButtonProps) => {
  switch (variant) {
    case 'primary': {
      return 'bg-green text-white hover:bg-green-light border-none';
    }
    case 'secondary': {
      return 'border-green-light text-green-light font-bold bg-transparent hover:bg-green-light hover:text-white';
    }
    case 'tertiary': {
      return 'bg-gray3 text-white hover:bg-gray4 border-none';
    }
  }
}

