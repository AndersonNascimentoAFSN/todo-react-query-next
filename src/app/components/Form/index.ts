import { Input } from './Input';
import { Label } from './Label';
import { Field } from './Field';
import { ErrorMessage } from './ErrorMessage';

export const Form = {
  Input,
  Label,
  Field,
  ErrorMessage,
}