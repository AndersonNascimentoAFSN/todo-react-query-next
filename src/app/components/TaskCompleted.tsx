'use client'

import { useTodoList } from "@/hooks/useTodoList";

export function TaskCompleted() {
  const { data: todos } = useTodoList();

  const taskCompleted = todos?.filter((todo) => todo.isCompleted).length

  return (
    <>
      <span className="text-white mr-3">Quantidade de tarefas completas:</span>
      <span className="text-white">{taskCompleted}</span>
    </>
  )
}
