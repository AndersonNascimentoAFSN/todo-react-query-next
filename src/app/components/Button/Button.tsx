import { useMemo } from "react"

import { BASE_BUTTON_CLASSES, getVariantClasses } from "./style";
import { ButtonProps } from "./type";

export function Button({ children, variant = "primary", ...rest }: ButtonProps) {
  const computedClasses = useMemo(() => {
    const variantClass = getVariantClasses(variant);

    return [variantClass].join(' ');
  }, [variant]);

  return (
    <button
      {...rest}
      className={`${BASE_BUTTON_CLASSES} ${computedClasses} ${rest?.className}`}
    >
      {children}
    </button>
  )
}

Button.displayName = 'Button'