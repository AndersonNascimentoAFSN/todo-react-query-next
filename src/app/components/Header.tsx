import { Suspense } from "react";
import { TaskCompleted } from "./TaskCompleted";
import HydratedTodos from "../hydratedTodos";

export function Header() {
  return (
    <header className="w-screen bg-gray1 p-10">
      <h1 className="m-auto text-white text-3xl">Lista de Tarefas</h1>
      <h3 className="m-auto text-white text-xl">Organize suas tarefas de forma simples e pr√°tica</h3>

      <Suspense fallback={<div>Loading...</div>}>
        {/* @ts-expect-error Server Component */}
        <HydratedTodos>
          <TaskCompleted />
        </HydratedTodos>
      </Suspense>
    </header>
  )
}
