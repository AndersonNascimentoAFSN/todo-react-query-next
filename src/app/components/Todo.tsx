'use client'

import { useCreateTodo } from "@/hooks/useCreateTodo"
import { useState } from "react"
import { Button } from "./Button"

export function Todo() {
  const [task, setTask] = useState<string>('')
  const { mutateAsync } = useCreateTodo()

  function handleAddTodo() {
    mutateAsync({ todo: { description: task } })
    setTask('')
  }

  return (
    <form className="w-screen p-10 flex gap-4">
      <input
        type="text"
        placeholder="adicione uma tarefa a lista"
        className="w-full max-w-lg p-2 border-2 border-gray-300 placeholder:text-purple-500 outline-none ring-1 focus:ring focus:ring-purple-300 text-purple-800"
        onChange={(event) => setTask(event.target.value)} value={task}
      />
      <Button
        onClick={handleAddTodo}
      >
        Adicionar
      </Button>
    </form>
  )
}