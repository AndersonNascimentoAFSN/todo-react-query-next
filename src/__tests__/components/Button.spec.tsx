import { render, screen, logRoles, act, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import { Button } from '@/app/components/Button'

describe('<Button />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should be able to render the button correctly', () => {
    render(<Button>Add</Button>)

    const buttonElement = screen.getByRole('button', { name: /add/i })

    expect(buttonElement).toBeVisible()
  })

  it('should be able to onClick has called', async () => {
    const handleAdd = jest.fn()

    const { container } = render(<Button onClick={handleAdd}>Add</Button>)

    const buttonElement = screen.getByRole('button', { name: /add/i })

    await act(async () => {
      await userEvent.click(buttonElement)
    })

    expect(handleAdd).toHaveBeenCalled()
    expect(handleAdd).toHaveBeenCalledTimes(1)

    expect(container).toMatchSnapshot()
  })

  it('should be able to onClick has called', async () => {
    render(<Button variant="primary">Add</Button>)

    const buttonElement = screen.getByRole('button', { name: /add/i })

  })
})