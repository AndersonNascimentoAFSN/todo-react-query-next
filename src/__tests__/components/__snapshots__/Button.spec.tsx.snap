// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should be able to onClick has called 1`] = `
<div>
  <button
    class="p-2 bg-purple-700 rounded-lg text-white hover:opacity-75 undefined"
  >
    Add
  </button>
</div>
`;
